### ISSUE TITLE
Add markers for special Unicode characters

### ISSUE THREAD

**Issue by vpetzel** [tonsky/FiraCode#1165]
Hello,
I suggest using some sort of marker for special Unicode chars such as thinspaces and other non standard spaces, zero width joiners and such things. I think in a coding font every char should be visible (so you know that they are there). I guess even a code such as ZWNJ for x200c for example would work.

Regards,
Valentin

**Comment by RubenVerg** [comment]
Wait, no? If I have a zwnj it's because I need to not join and for it to be *zero width*

If you have a strange use-case when you need to see them, there surely are extensions for your favorite IDEs.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by vpetzel** [comment]
Not really. As I understand it, this font should be tuned for Code, and usually Code is not what you want, but a representation of what you want. Usually you want a zwnj for output and Typography, but not in Code. But the Code should be able to give all informations about what’s there, and should definitely not have any invisible characters. Source code should always be WYSIWYM, not WYSIWYG.

We could for example just use an opentype character variation to have such characters visible on demand.

**Comment by alexeyten** [comment]
Nice idea.
Actually, long time ago I patched some fonts to show `NBSP` as underline with small letters `sp` under it. And em-dash/en-dash with letters `m` and `n` over dash.

**Comment by RubenVerg** [comment]
Generally so, but some characters are specially defined as being invisible and "control", would you like to see "LF" instead of a newline? Also lol am I the only one using Fira Code for UI?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by RubenVerg** [comment]
No what? How can you work with an em dash not looking like a dash, but like a strange underlined small M?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by vpetzel** [comment]
> Generally so, but some characters are specially defined as being invisible and "control", would you like to see "LF" instead of a newline? Also lol am I the only one using Fira Code for UI? Get Outlook for Android<https://aka.ms/ghei36>
> […](#)
> ________________________________ From: Valentin Petzel <notifications@github.com> Sent: Monday, December 7, 2020 10:23:06 AM To: tonsky/FiraCode <FiraCode@noreply.github.com> Cc: Ruben Vergani <RubenVerg@outlook.com>; Comment <comment@noreply.github.com> Subject: Re: [tonsky/FiraCode] Add markers for special Unicode characters (#1165) Not really. As I understand it, this font should be tuned for Code, and usually Code is not what you want, but a representation of what you want. Usually you want a zwnj for output and Typography, but not in Code. But the Code should be able to give all informations about what’s there, and should definitely not have any invisible characters. Source code should always be WYSIWYM, not WYSIWYG. We could for example just use an opentype character variation to have such characters visible on demand. — You are receiving this because you commented. Reply to this email directly, view it on GitHub<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftonsky%2FFiraCode%2Fissues%2F1165%23issuecomment-739790389&data=04%7C01%7C%7Ca2314be153a04d6329b908d89a91b51f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637429297877407573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=yXOCFTRpljVWMHqzpnqCsJ19T3OpiMY1OKtZHP42R%2FI%3D&reserved=0>, or unsubscribe<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMRJI3AYAM6JXSE2DBUKYC3STSNHVANCNFSM4UPVQE4A&data=04%7C01%7C%7Ca2314be153a04d6329b908d89a91b51f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637429297877407573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=8YBqAH%2BX%2FonFhowrQL6kjat4MdhibQ5hwewORFhvUUQ%3D&reserved=0>.

But you do see LF in the form of a new line. The thing is: most control characters have absolutely no reason to appear in source code unless you want to specify that particular character. And in these cases it is useful for the characters to be actually seen, so you know they are there. E.g. let’s say im writing a text using a text file based processor, I’d want to prevent the ff-Ligature in ”Aufführungspraxis“ by adding a zwnj in between, making it ”Auf‌führungspraxis“. This will look exactly the same in a Text Editor, so it is almost impossible to keep track of where these things are. If we had something like ”Auf~führungspraxis“ where ~ is a placeholder for some sort of marking this would be much easier to keep track of. Or let’s say you want to have something like ”7. 12. 2020“. Here you would want thinspaces in between, so ”7. 12. 2020“. In a standard monospaced font there is absolutely no optical distinction between these two, making it again very hard to keep track of such things.

I don’t know what benefits FiraCode should have over Fira Mono for UI, but as I said, simply making use of opentype character variations will enable us to have both on demand.

**Comment by RubenVerg** [comment]
The benefits are ligatures, obviously

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by tonsky** [comment]
I suspect rendering ZWNJ might break e.g. emoji

**Comment by vpetzel** [comment]
> I suspect rendering ZWNJ might break e.g. emoji

I don’t think it would. An emoji ligature does not require a zwnj, generally you’d only need the zwnj to prevent the emoji ligature from happening. But I think control characters like zwnj require software support, as software will usually not print these characters.

**Comment by RubenVerg** [comment]
I suppose they meant rendering ZWJ would break
________________________________
Da: Valentin Petzel <notifications@github.com>
Inviato: lunedì 7 dicembre 2020 15:14
A: tonsky/FiraCode <FiraCode@noreply.github.com>
Cc: Ruben Vergani <RubenVerg@outlook.com>; Comment <comment@noreply.github.com>
Oggetto: Re: [tonsky/FiraCode] Add markers for special Unicode characters (#1165)


I suspect rendering ZWNJ might break e.g. emoji

I don’t think it would. An emoji ligature does not require a zwnj, generally you’d only need the zwnj to prevent the emoji ligature from happening. But I think control characters like zwnj require software support, as software will usually not print these characters.
